### PR TITLE
Included linear + binary search for 1D static arrays

### PR DESCRIPTION
**Pull request by skullVoid**
#### References to other Issues or PRs or Relevant literature

#446 

#### Brief description of what is fixed or changed
A search function is added. 
It has 2 necessary parameters(1. element to search, 2. order of the array) and one optional (3. comparator)
1. what to search
2. positive values (int or float allowed, anything else raises type error exception) mean ascending order, negative is descending, 0 is unsorted
3. function as a parameter; this is for when the usual >, < do not work on the datatype of the array

### CODE REVIEW

File: pydatastructs/linear_data_structures/arrays.py [hunk 1]
```diff
@@ -126,7 +126,7 @@ def __new__(cls, dtype=NoneType, *args, **kwargs):
     @classmethod
     def methods(cls):
         return ['__new__', '__getitem__',
-        '__setitem__', 'fill', '__len__']
+        '__setitem__', 'fill', '__len__', 'search']
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/arrays.py:129]: Please make it a function (like place in `algorithms.py` in `linear_data_structures`) instead of a class method?

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: Pardon me, shall I include this directly? Since all this does is search in 1D arrays. I wanted to include this for dynamic arrays too, but that removal of items just makes a void location and does not shift the items in those void locations

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: Regardless, I will include a jump search as it was mentioned as an idea in the issue. Again, for 1D static arrays. Is that fine?

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/arrays.py:129]: That should be fine. I am planning to add `ImplicitArray` so it would be useful to have search functions for that as well. And anyways, conceptually, search is an algorithm that just reads the data from an array. It's not responsible for updating the state of `OneDimensionalArray` class, so having it as a method is not of much use.

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: Understood. So should I just move this change to algorithms.py along with addition of the jump search?

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/arrays.py:129]: Yes. Though let's focus only on one search algorithm for now. I think you are adding `binary_search`. Please make it that clear while naming the function. Like explicitly name the functions as `binary_search`, `jump_search`, `ternary_search`, etc.

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: Actually, it is a hybrid. The search depends on the sorted order parameter. Positive means ascending sorted, zero is unsorted and negative is descending sorted.

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: For 0 the search is simply linear, else it is binary. There is only a change of the comparison sign for descending array

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/arrays.py:129]: I don't think that's a good idea. The logic for each type of array should be encapsulated in a single function. The user can decide on their own what they might want to use. Please make separate functions for `binary_search`, `linear_search`, etc.

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: If you prefer that I should separate this hybrid function, then I shall do that

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: Ok understood

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:129]: Uh sorry to ask this. Should I make a new PR after doing as told?

Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py]: I pushed the new commit.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 2]
```diff
@@ -1369,3 +1372,58 @@ def bubble_sort(array, **kwargs):
         array._modify(force=True)
 
     return array
+
+def linear_search(array, elem):
+    for i in range(array._size):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1377]: Is `array.size` not working? We shouldn't access private attributes (the ones having `_` as prefix) unless really needed.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 3]
```diff
@@ -1369,3 +1372,58 @@ def bubble_sort(array, **kwargs):
...
+        if array._data[i] == elem:
+            return i
+    return -1
+
+def binary_search(array, elem, comp = None):
+    L = 0
+    R = array._size - 1
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1384]: `left`, `right`?

File: pydatastructs/linear_data_structures/arrays.py [hunk 4]
```diff
@@ -148,6 +148,7 @@ def fill(self, elem):
 
     def __len__(self):
         return self._size
+
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/arrays.py:151]:
```suggestion
```

**Review by czgdp1807 — DISMISSED** [review]

File: pydatastructs/linear_data_structures/algorithms.py (hunk 2 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1377]: no no. While seeing the code, I noticed that only _size is being used, so I used that too

File: pydatastructs/linear_data_structures/algorithms.py (hunk 3 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1384]: Yes

File: pydatastructs/linear_data_structures/arrays.py (hunk 4 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/arrays.py:151]: Sorry, I didnt get this

File: pydatastructs/linear_data_structures/algorithms.py [hunk 5]
```diff
@@ -1369,3 +1372,58 @@ def bubble_sort(array, **kwargs):
...
+            else:
+                R = M - 1
+        else:
+            if comp(array._data[M], elem):
+                L = M + 1
+            else:
+                R = M - 1
+    return -1
+
+def jump_search(array, elem, comp = None):
```
[the remark this answers is not in the document]
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1401]: No, I am afraid of making mistakes since I dont have much expertise, so I did the way I was confident in.

[the remark this answers is not in the document]
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1401]: Sorry for coming short

[the remark this answers is not in the document]
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1401]: Are there any issues?

File: pydatastructs/linear_data_structures/algorithms.py (hunk 3 above)
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:1384]: Please make it explicit then. Always use full names instead of short hand notations. Not many key strokes will be saved while typing, `L` instead of `left`.

File: pydatastructs/linear_data_structures/arrays.py (hunk 4 above)
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/arrays.py:151]: It's a suggestion to remove that extra line. It's not needed.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:1401]: Yes. All the APIs should be consistent to the maximum extent possible. Now, here, `start`, `end` and `backend` should be included and the implementation should be modified to use these parameters.

[the remark this answers is not in the document]
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1401]: There is some git issue I cannot understand, I have created a new PR

[the remark this answers is not in the document]
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py]: Well, you had to first pull the changes from `forArrays` branch. `git pull your_remote_name forArrays`. That will fetch the suggestion you applied here, then you will be able to push to your `forArrays` branch.

[the remark this answers is not in the document]
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py]: Is there still some issue? Since this is marked as a draft.

[the remark this answers is not in the document]
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py]: Please add tests for all your APIs. See, `tests/test_algorithms.py` file under, `linear_data_structures`.

[the remark this answers is not in the document]
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py]: Doc string are also to be added. See other algorithms in the same file.

[the remark this answers is not in the document]
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py]: On it

File: pydatastructs/linear_data_structures/algorithms.py [hunk 6]
```diff
@@ -1369,3 +1372,197 @@ def bubble_sort(array, **kwargs):
...
+        If not found, returns -1.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import OneDimensionalArray, linear_search
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> out = linear_search(arr, 2)
+    >>> print(out)
+    1
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1414]:
```suggestion
    >>> linear_search(arr, 2)
    1
```
Above should work. Can you try?

File: pydatastructs/linear_data_structures/algorithms.py [hunk 7]
```diff
@@ -1369,3 +1372,197 @@ def bubble_sort(array, **kwargs):
...
+    """
+    raise_if_backend_is_not_python(
+            binary_search, kwargs.get('backend', Backend.PYTHON))
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+    Left = start
+    Right = end
+    while Left <= Right:
+        Middle = int((Left+Right)/2)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py:1490]: What about, `Left//2 + Right//2`?

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 8]
```diff
@@ -322,3 +322,24 @@ def test_next_prev_permutation():
...
+    array = ODA(int, [1, 2, 5, 7, 10, 29, 40])
+    for i in range(len(array)):
+        assert i == binary_search(array, array[i])
+
+def test_jump_search():
+    ODA = OneDimensionalArray
+
+    array = ODA(int, [1, 2, 5, 7, 10, 29, 40])
+    for i in range(len(array)):
+        assert i == jump_search(array, array[i])
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:345]: What about corner cases like, empty array, one element array? And see `_test_common_sort` above. Repetition of tests can be avoided by doing something similar here.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 7 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1490]: I think (Left + Right)//2 is fine but not too sure about the sum of floors, for example, when left = 3, right = 5(when both odd), it will give 3 as middle

File: pydatastructs/linear_data_structures/tests/test_algorithms.py (hunk 8 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/tests/test_algorithms.py:345]: Are you asking for the test function of the actual algorithm written?

File: pydatastructs/linear_data_structures/algorithms.py (hunk 6 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1414]: Yes Sure

Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1414]: I did not test it in conda environment so i am unfamiliar. But it works as intended, I checked in vs code.

File: pydatastructs/linear_data_structures/algorithms.py (hunk 7 above)
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:1490]: Well the loss won't be much. There are four cases,

1. Left - Odd, Right - Odd - Middle will be lesser (by 1) in Left//2 + Right//2 as compared to current approach.
2. Left - Odd, Right - Even or Right - Odd, Left - Even - Middle will be exactly the same as in current approach.
3. Left - Even, Right - Even - Middle will again be exactly the same as in current approach.

The reason is (Left + Right) might get too large for implicit arrays leading to slow down in performance (Python uses big integers for handling larger than 64 bit integers). So, basically if you can just add 1 for Left - Odd, Right - Odd in Left//2 + Right//2 and do only Left//2 + Right//2 for other cases then it should be fine. In fact,

`Middle = Left//2 + Right//2 + Left % 2 * Right % 2` should work. Worth trying.

Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1490]: Yes, this seems much better, I shall do this

File: pydatastructs/linear_data_structures/tests/test_algorithms.py (hunk 8 above)
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/tests/test_algorithms.py:345]: Yes. You are using the same test for every search algorithm. So why not write a common test function, `_test_common_search(search_function_api)` and then in `test_jump_search` (for example),

```python
def test_jump_search():
    _test_common_search(jump_search)
```

**Comment by czgdp1807** [comment]
See [this](https://app.codecov.io/gh/codezonediitj/pydatastructs/compare/454/diff). The read lines are not covered. Add tests for covering them as well.

File: pydatastructs/linear_data_structures/tests/test_algorithms.py (hunk 8 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/tests/test_algorithms.py:345]: Right, I will do this

File: pydatastructs/linear_data_structures/algorithms.py (hunk 6 above)
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:1414]: I think only Python >= 3.5 is needed for testing. Doesn't matter whether it comes from conda, or VS Code.

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py:1414]: See the testing section of README.

Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1414]: Yes, I havent done it like that, I just ran sample codes to check. But it works

**Comment by skullVoid** [comment]
> See [this](https://app.codecov.io/gh/codezonediitj/pydatastructs/compare/454/diff). The red lines are not covered. Add tests for covering them as well.

Its showing that bubble sort is not covered, but it is, I did not touch anything other than searching which I added

File: pydatastructs/linear_data_structures/algorithms.py (hunk 6 above)
Comment (skullVoid) [reply pydatastructs/linear_data_structures/algorithms.py:1414]: Ok, will do

**Comment by czgdp1807** [comment]
> Its showing that bubble sort is not covered, but it is, I did not touch anything other than searching which I added

That's the red (lighter shade) of diff. If you scroll and click on `Click to load diff` then you will see dark red lines. These are the ones that you have to cover by adding tests.

**Comment by skullVoid** [comment]
> > Its showing that bubble sort is not covered, but it is, I did not touch anything other than searching which I added
> 
> That's the red (lighter shade) of diff. If you scroll and click on `Click to load diff` then you will see dark red lines. These are the ones that you have to cover by adding tests.

Ok on it. But still, bubble sort and another line of importing is in deep red, what about it?

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 9]
```diff
@@ -323,23 +323,22 @@ def test_next_prev_permutation():
...
+        assert i == search(array, array[i])
+    
+    checker_array = [-1, -1, 2, 3, 4, 5, -1]
+    for i in range(len(array)):
+        assert checker_array[i] == search(array, array[i], start=2, end=5)
 
-def test_binary_search():
-    ODA = OneDimensionalArray
+def test_linear_search():
+    _test_common_search(test_linear_search)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:338]:
```suggestion
    _test_common_search(linear_search)
```

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 10]
```diff
@@ -323,23 +323,22 @@ def test_next_prev_permutation():
...
 
-    array = ODA(int, [1, 2, 5, 7, 10, 29, 40])
-    for i in range(len(array)):
-        assert i == binary_search(array, array[i])
+def test_binary_search():
+    _test_common_search(test_binary_search)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:341]:
```suggestion
    _test_common_search(binary_search)
```

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 11]
```diff
@@ -323,23 +323,22 @@ def test_next_prev_permutation():
...
 
 def test_jump_search():
-    ODA = OneDimensionalArray
-
-    array = ODA(int, [1, 2, 5, 7, 10, 29, 40])
-    for i in range(len(array)):
-        assert i == jump_search(array, array[i])
+    _test_common_search(test_jump_search)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:344]:
```suggestion
    _test_common_search(jump_search)
```

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 12]
```diff
@@ -323,23 +323,22 @@ def test_next_prev_permutation():
         _, orig_array = next_permutation(prev_array)
         assert str(orig_array) == str(array)
 
-def test_linear_search():
+def _test_common_search(search):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:326]:
```suggestion
def _test_common_search(search_func):
```

File: pydatastructs/linear_data_structures/tests/test_algorithms.py [hunk 13]
```diff
@@ -323,23 +323,22 @@ def test_next_prev_permutation():
...
     ODA = OneDimensionalArray
 
     array = ODA(int, [1, 2, 5, 7, 10, 29, 40])
     for i in range(len(array)):
-        assert i == linear_search(array, array[i])
+        assert i == search(array, array[i])
+    
+    checker_array = [-1, -1, 2, 3, 4, 5, -1]
+    for i in range(len(array)):
+        assert checker_array[i] == search(array, array[i], start=2, end=5)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/tests/test_algorithms.py:335]:
```suggestion
        assert checker_array[i] == search_func(array, array[i], start=2, end=5)
```

**Comment by czgdp1807** [comment]
> bubble sort and another line of importing is in deep red, what about it?

Ignore. Though you can generate the same report locally if Codecov UI is confusing. See the testing section of README. Once you execute the command there, you will have to open the `index.html` file in `html/cov` in a browser.

**Comment by skullVoid** [comment]
> > bubble sort and another line of importing is in deep red, what about it?
> 
> Ignore. Though you can generate the same report locally if Codecov UI is confusing. See the testing section of README. Once you execute the command there, you will have to open the `index.html` file in `html/cov` in a browser.

Can I use Codecov?

**Comment by czgdp1807** [comment]
Sure.

**Comment by skullVoid** [comment]
> Sure.

Is there any other issue?

**Comment by czgdp1807** [comment]
I will push some changes may be on Sunday. I don't think there is anything else to be done from your side.

**Comment by skullVoid** [comment]
> I will push some changes to your branch `forArrays` may be on Sunday. I don't think there is anything else to be done from your side.

Please let me know if there is any error. Shall I see if I can contribute something else in the project?

**Comment by czgdp1807** [comment]
> Shall I see if I can contribute something else in the project?

Sure go ahead.

**Comment by czgdp1807** [comment]
Thanks @skullVoid

**Comment by skullVoid** [comment]
> Thanks @skullVoid

I didnt do much. Thank you for guiding me, this was a good experience for me as a first timer.
